### ISSUE TITLE
VSTests throws a FileNotFoundException in project to project references.

### ISSUE THREAD

**Issue by alexdrl** [Microsoft/vstest#1480]
## Description
VSTests throws a FileNotFoundException when trying to launch related tests DLL when a test project references another test project (.NET Core).

## Steps to reproduce
Having a solution with 2 test projects, one of them referencing the other.

## Expected behavior
VSTests should execute the tests in its expected folder and ignore the tests DLLs that are copied over by the compilation.

## Actual behavior
VSTest throws an exception (FileNotFoundException) when loading a test DLL that is not located in its compilation folder. As you can see below, VSTests throws in Application.Tests, but in Proxy.Tests folder.

## Diagnostic logs
```
2018-03-15T09:29:40.6984718Z System.IO.FileNotFoundException: No test is available in D:\Agents\2\_work\57\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\A.R.H.Application.Tests.dll. Make sure test project has a nuget reference of package "Microsoft.NET.Test.Sdk" and framework version settings are appropriate and try again.
2018-03-15T09:29:40.6984718Z    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Hosting.DotnetTestHostManager.GetTestHostProcessStartInfo(IEnumerable`1 sources, IDictionary`2 environmentVariables, TestRunnerConnectionInfo connectionInfo)
2018-03-15T09:29:40.6984718Z    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, CancellationToken cancellationToken)
2018-03-15T09:29:40.6984718Z    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyExecutionManager.StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler eventHandler)

```
## Environment
Tests launched by a TFS 2018 VSTest task, in a VS 15.6 build agent.
```
2018-03-15T09:29:32.4016478Z ========================================================
2018-03-15T09:29:32.4016478Z Test selector : Test assemblies
2018-03-15T09:29:32.4016478Z Test assemblies : **\release\**\*.tests*.dll,!**\obj\**
2018-03-15T09:29:32.4016478Z Test filter criteria : null
2018-03-15T09:29:32.4016478Z Search folder : D:\Agents\2\_work\57\s
2018-03-15T09:29:32.4016478Z Run settings file : D:\Agents\2\_work\57\s
2018-03-15T09:29:32.4016478Z Run in parallel : false
2018-03-15T09:29:32.4016478Z Run in isolation : false
2018-03-15T09:29:32.4016478Z Path to custom adapters : null
2018-03-15T09:29:32.4016478Z Other console options : null
2018-03-15T09:29:32.4016478Z Code coverage enabled : true
2018-03-15T09:29:32.4016478Z VisualStudio version selected for test execution : latest
2018-03-15T09:29:32.4016478Z [command]D:\Agents\2\_work\_tasks\VSTest_ef087383-ee5e-42c7-9a53-ab56c98420f9\2.1.14\vswhere.exe -version [15.0,16.0) -latest -products * -requires Microsoft.VisualStudio.PackageGroup.TestTools.Core -property installationPath
2018-03-15T09:29:32.4016478Z C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
2018-03-15T09:29:32.4016478Z [command]C:\windows\System32\Wbem\WMIC.exe datafile where "name='C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Enterprise\\Common7\\IDE\\CommonExtensions\\Microsoft\\TestWindow\\vstest.console.exe'" get Version /Value
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z Version=15.0.27323.2
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4172739Z 
2018-03-15T09:29:32.4641482Z ========================================================
2018-03-15T09:29:33.7141399Z [command]"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TestWindow\vstest.console.exe" D:\Agents\2\_work\57\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\A.R.H.Application.Tests.dll D:\Agents\2\_work\57\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\A.R.H.Application.Tests.dll D:\Agents\2\_work\57\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\A.R.H.Proxy.Tests.dll /EnableCodeCoverage /logger:trx "/TestAdapterPath:\"D:\Agents\2\_work\57\s\""
2018-03-15T09:29:33.8547638Z Microsoft (R) Test Execution Command Line Tool Version 15.6.0
2018-03-15T09:29:33.8547638Z Copyright (c) Microsoft Corporation.  All rights reserved.
2018-03-15T09:29:33.8547638Z 
2018-03-15T09:29:34.0891373Z Starting test execution, please wait...

```

**Comment by smadala** [comment]
- Are these .NET Core tests assemblies(A.R.H.Application.Tests.dll) build on same machine? If not, do `dotnet publish` on A.R.H.Application.Tests.csproj, copy artifact this machine and run tests.

- A.R.H.Application.Tests csproj has reference to "Microsoft.NET.Test.Sdk" package? 
- Can you share csproj of tests project?

**Comment by alexdrl** [comment]
> Are these .NET Core tests assemblies(A.R.H.Application.Tests.dll) build on same machine? If not, do dotnet publish on A.R.H.Application.Tests.csproj, copy artifact this machine and run tests.

Yes, they are built on the same VS solution.

> A.R.H.Application.Tests csproj has reference to "Microsoft.NET.Test.Sdk" package?

Yes

> Can you share csproj of tests project?

https://gist.github.com/alexdrl/016e16f895eaaffa7fd88cd2c0c6f657

**Comment by smadala** [comment]
Can you queue a build with `system.debug` and `agent.diagnostic` variables to true and share build logs?
 `
![image](https://user-images.githubusercontent.com/4337332/37789183-439eb1b2-2e29-11e8-9f54-c1b527d364b0.png)

**Comment by alexdrl** [comment]
Those are the logs: [logs_15478.zip](https://github.com/Microsoft/vstest/files/1840542/logs_15478.zip)

**Comment by smadala** [comment]
To run tests for .NET core you need test-assembly's runtimeconfig.dev.json, runtimeconfig.json and deps.json files. These files are not present with `ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.dll`.
Can you find why, What is special about this project?

Are you using fakes in .NET Core test (Fakes not supported for .NET Core)? 

Are you able to run these tests locally(on your dev box)?

```
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.dll (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.pdb (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.deps.json (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.dll (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.pdb (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.runtimeconfig.dev.json (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.runtimeconfig.json (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.xml (file)
2018-03-23T07:26:57.7863759Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Application\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.xml (file)



2018-03-23T07:26:57.1457559Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.dll (file)
2018-03-23T07:26:57.1457559Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.pdb (file)
2018-03-23T07:26:57.1457559Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.dll (file)
2018-03-23T07:26:57.1457559Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.pdb (file)
2018-03-23T07:26:57.1457559Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.Tests.xml (file)
2018-03-23T07:26:57.1457559Z ##[debug]  D:\Agents\1\_work\69\s\DslTools\Application\Debugging\Proxy\Tests\bin\Release\netcoreapp2.0\ADDInformatica.RPQPrototype.HealthAndWelfare.Application.xml (file)

```

**Comment by alexdrl** [comment]
> Can you find why, What is special about this project?

I don't know, but they are similar projects.

> Are you using fakes in .NET Core test (Fakes not supported for .NET Core)?

The only framework library used is Moq

> Are you able to run these tests locally(on your dev box)?

Yes, using Visual Studio and `dotnet test`

**Comment by smadala** [comment]
You can use `.NET Core` task to build and test. If you can repro the same issue with vstest.console.exe locally, share the sample project.

**Comment by alexdrl** [comment]
Here is the sample project. Two test projects and a file that contains the vstest.console.exe command similar to which the TFS task launches.

[TestProjectA.zip](https://github.com/Microsoft/vstest/files/1847336/TestProjectA.zip)

**Comment by smadala** [comment]
Thanks for repro. I got the problem now. You are trying to run tests of `TestProjectA.dll` two times. 
- TestProjectA.dll from TestProjectA project output directory. this works fine because deps.json, runtimeconfig.dev.json and runtimeconfig.json files exits with it.
- TestProjectA.dll from TestProjectB project output directory. this fails because not able to find above mentioned files.

In `dotnet test` case it run tests from TestProjectA only once.

I believe your intention not to run tests twice(TestProjectA.dll coming twice in command due to patterns  in `Test assemblies` in vstest task). You might be using vstest task for code coverage.

I suggest following changes to your build definition to overcome this issue.
- Use `dotnet publish` to build your project. Something like `dotnet publish -o C:\Users\samadala\DRI\TestProjectA\artifacts`.
- Now you have all the test assemblies and required *.json in `C:\Users\samadala\DRI\TestProjectA\artifacts`, you can update the `Test assemblies` value accordingly.

**Comment by alexdrl** [comment]
> I believe your intention not to run tests twice(TestProjectA.dll coming twice in command due to patterns in Test assemblies in vstest task). You might be using vstest task for code coverage.

Yes, you are 100% correct. 

We can not use `dotnet publish` to build the tests as the build requires VS for StrongNamer. We could use `dotnet publish`, but... This implies that every time a test is added to the solution, this task should be updated?

If `dotnet test` supported test coverage, we could use it, and the problem will go away for us, but not for VSTest users...

**Comment by smadala** [comment]
>We can not use dotnet publish to build the tests as the build requires VS for StrongNamer. We could use dotnet publish, but... This implies that every time a test is added to the solution, this task should be updated?

Following are options.

- You can avoid running tests of `TestProjectA` twice, by tuning `Test assemblies` option(Some regex work. 😄).
- Use `dotnet publish` at solution level. This makes no need update the task whenever you add new test project/test.

>If dotnet test supported test coverage, we could use it, and the problem will go away for us, but not for VSTest users...

Code Coverage support with `dotnet test` is coming soon.

**Comment by smadala** [comment]
@alexdrl  Above mentioned options should work for you. 

/cc @cltshivash FYI.

I'm closing the issues as by design and won't fix.

**Comment by JamesKibirige** [comment]
Hi, @smadala I am having a similar problem running my integration and acceptance tests in my release pipeline I see this same error. However I have already made sure that I have published all the test binaries and dependencies for the Test projects as Build artefacts, they are in folders next to the Test dlls, but I still see this problem:
```
2019-01-16T21:18:14.5234936Z vstest.console.exe 
2019-01-16T21:18:14.5235596Z "D:\a\r1\a\drop\Tests.Integration\Tests.Integration.dll"
2019-01-16T21:18:14.5235789Z /logger:"trx"
2019-01-16T21:18:14.5235976Z /TestAdapterPath:"D:\a\r1\a\drop\Tests.Integration"
2019-01-16T21:18:19.6372101Z Starting test execution, please wait...
2019-01-16T21:18:24.2975659Z System.IO.FileNotFoundException: Unable to find tests for D:\a\r1\a\drop\Tests.Integration\Tests.Integration.dll. Make sure test project has a nuget reference of package "Microsoft.NET.Test.Sdk" and framework version settings are appropriate. Rerun with /diag option to diagnose further.
2019-01-16T21:18:24.2977430Z    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Hosting.DotnetTestHostManager.GetTestHostProcessStartInfo(IEnumerable`1 sources, IDictionary`2 environmentVariables, TestRunnerConnectionInfo connectionInfo)
2019-01-16T21:18:24.2977592Z    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources)
2019-01-16T21:18:24.2977657Z    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyExecutionManager.StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler eventHandler)
```
